client/allochealth/tracker.go

```diff
@@ -379,7 +381,12 @@ OUTER:
 				allocReg = newAllocReg
 			}
 		case <-healthyTimer.C:
-			t.setCheckHealth(true)
+			if t.setCheckHealth(true) {
+				// final health set and propagated
+				return
```

Comment:
The docstring for this function says this is a "long lived goroutine". If we return here, how does the tracker detect service check failures that happen later on?

Reply:
"long lived goroutine" is a bit misnomer - it runs until the alloc is marked healthy or terminally unhealthy.  Once an alloc is marked healthy, it never marks it unhealthy.

The exiting goroutine happens today already.  `setCheckHealth` and `setTaskHealth` call `t.cancelFn()` which will cause both `watchTaskEvents` and `watchConsulEvents` as they handle `t.ctx.Done()`.

Reply:
Ah, gotcha.